samcli/lib/bootstrap/companion_stack/companion_stack_manager.py

```diff
@@ -0,0 +1,259 @@
+"""
+    Companion stack manager
+"""
+from typing import List, Dict
+import typing
+
+import boto3
+from botocore.config import Config
+from botocore.exceptions import ClientError, NoRegionError, NoCredentialsError
+
+from samcli.commands.exceptions import CredentialsError, RegionError
+from samcli.lib.bootstrap.companion_stack.companion_stack_builder import CompanionStackBuilder
+from samcli.lib.bootstrap.companion_stack.data_types import CompanionStack, ECRRepo
+from samcli.lib.package.artifact_exporter import mktempfile
+from samcli.lib.package.s3_uploader import S3Uploader
+
+# pylint: disable=E0401
+if typing.TYPE_CHECKING:  # pragma: no cover
+    from mypy_boto3_cloudformation.client import CloudFormationClient
+    from mypy_boto3_s3.client import S3Client
+else:  # pragma: no cover
+    CloudFormationClient = object
+    S3Client = object
+
+
+class CompanionStackManager:
+    """
+    Manager class for a companion stack
+    Used to create/update the remote stack
+    """
+
+    _companion_stack: CompanionStack
+    _builder: CompanionStackBuilder
+    _boto_config: Config
+    _s3_bucket: str
+    _s3_prefix: str
+    _cfn_client: CloudFormationClient
+    _s3_client: S3Client
+
+    def __init__(self, stack_name, region, s3_bucket, s3_prefix):
+        self._companion_stack = CompanionStack(stack_name)
+        self._builder = CompanionStackBuilder(self._companion_stack)
+        self._boto_config = Config(region_name=region if region else None)
+        self._s3_bucket = s3_bucket
+        self._s3_prefix = s3_prefix
+        try:
+            self._cfn_client = boto3.client("cloudformation", config=self._boto_config)
+            self._ecr_client = boto3.client("ecr", config=self._boto_config)
+            self._s3_client = boto3.client("s3", config=self._boto_config)
+            self._account_id = boto3.client("sts").get_caller_identity().get("Account")
```

Comment:
This is interesting, we are now using additional permissions, do we really need the account id? I see its used for fetching the ecr uri, is there no direct api call that will help us there?

Customers may or may not have sts permissions, especially as this does not directly map with the resources they are trying to deploy.

Reply:
GetCallerIdentity requires no permission and it's the recommended way of getting account ID.
https://docs.aws.amazon.com/IAM/latest/UserGuide/console_account-alias.html
https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html